src/vg_active_segment.erl

```diff
@@ -99,67 +101,127 @@ init([Topic, Partition, NextNode, Tab]) ->
+                history_tab = Tab
                }}.
 
-handle_call({write, RecordSet}, _From, State=#state{id=Id,
-                                                    topic=Topic,
-                                                    partition=Partition,
-                                                    next_brick=NextBrick}) ->
-    State1 = write_record_set(RecordSet, State),
-    case NextBrick of
-        {_, solo} -> ok;
-        {_, tail} -> ok;
-        {_, last} -> ok;
-        _ ->
-            gen_server:cast(NextBrick, {write, self(), RecordSet})
-    end,
-    LatestId = State1#state.id,
-    %% Should we do this in ack instead if it isn't solo or tail?
-    vg_topics:update_hwm(Topic, Partition, LatestId-1),
-    {reply, {ok, Id}, State1};
+handle_call({write, ExpectedID0, RecordSet}, _From, State = #state{id = ID,
+                                                                   topic = Topic,
+                                                                   partition = Partition,
+                                                                   next_brick = NextBrick}) ->
+    %% TODO: add pipelining of requests
+    try
+        ExpectedID =
+            case ExpectedID0 of
+                head ->
+                    ID + length(RecordSet);
+                Supplied when is_integer(Supplied) ->
+                    case ID + length(RecordSet) == Supplied of
+                        true ->
+                            ExpectedID0;
+                        %% should we check > vs < here?  one is repair
+                        %% the other is bad corruption
+                        _ ->
+                            %% inferred current id of the writing segment
+                            WriterID = ExpectedID0 - length(RecordSet),
+                            %% this should probably be limited, if
+                            %% we're going back too far, we need to be
+                            %% in some sort of catch-up mode
+                            lager:info("starting write repair, ~p", [WriterID]),
+                            WriteRepairSet = write_repair(WriterID, State),
+                            throw({write_repair, WriteRepairSet, State})
+                    end
+            end,
+        lager:info("write call from ~p", [_From]),
+
+        Result =
+            case NextBrick of
+                {_, Role} when Role == solo; Role == tail -> proceed;
+                _ ->
+                    (fun Loop(_, _, Remaining) when Remaining =< 0 ->
+                             {error, timeout};
+                         Loop(B, Start, Remaining) ->
+                             {Time, B1} = backoff:fail(B),
+                             case catch gen_server:call(NextBrick,
+                                                        {write, ExpectedID, RecordSet},
+                                                        Remaining) of
+                                 %% {error, timeout} ->
```

Comment:
should this commented out part be removed?

Reply:
yeah